Add documentation about the Qiskit Code Assistant

cbjuan opened the pull request:
1st draft for documentation about the Qiskit Code Assistant. Should fix #1930 

Considerations:
- Rough draft, feel free to edit or give any feedback
- I'm not sure if we should split the documentation in 2 parts: one for JupyterLab extension and another one for the VSCode extension
- I'm not sure the where the Qiskit Code Assistant should be placed in the TOC
- I added content that depends on #1918

javabster commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,230 @@
+
+# Qiskit Code Assistant
+
+Improve your quantum computing developer experience and learn best practices for Qiskit and IBM Quantum Platform services with Qiskit Code Assistant!
```
This first paragraph sounds a bit like a marketing advert, could you adjust the tone a bit to be more documentation style?

javabster commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,230 @@
+
+### Authentication and setup
+
+After installing the extension in VSCode or JupyterLab, the extension will try to authenticate you. By default, the package tries to authenticate to IBM Quantum services with the defined Qiskit API token, and uses your token from the `QISKIT_IBM_TOKEN` environment variable or from the file `~/.qiskit/qiskit-ibm.json` (under the section `default-ibm-quantum`).
```
do we need to explain to people how to set their api token env var?

javabster commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,230 @@
+
+Jupyterlab also includes a traditional completer than displays suggestion in a context menu rather than inline. This completer can be triggered using `Tab` to run and display the context menu.
+
+The context menu will include suggestions from the default completer in addition to the Qiskit Code Assistant suggestions. The context menu also sanitizes and trims the suggestions, making it less useful for see the code suggestion before inserting it.
```
typos?

javabster reviewed (commented):
Generally looks good! Just a few comments.

Question: do you think it would be better to split the content into 2 pages? one for the vscode extension and one for the jupyter lab extension? I felt like the page is quite long while i was reading and if a reader is only interested in one of the methods theres a lot of redundant info

beckykd commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,230 @@
+1. Launch Visual Studio Code
+1. Click on the Extensions tab
+1. Search for `qiskit-vscode`
+1. Click **Install** for the "Qiskit Code Assistant"
```
```suggestion
1. Find "Qiskit Code Assistant" and click **Install**.
```

beckykd reviewed (commented):
Thanks for this draft.  I'll finish my review tomorrow!

frankharkins commented:
> ...do you think it would be better to split the content into 2 pages? one for the vscode extension and one for the jupyter lab extension?

I think the tabs component could work quite well here

cbjuan commented:
> > ...do you think it would be better to split the content into 2 pages? one for the vscode extension and one for the jupyter lab extension?
> 
> I think the tabs component could work quite well here

Splitting in tabs

cbjuan commented:
Now the guide uses tabs: one giving an overview about the assistant, other two tabs for the extensions. Addressed some other comments and added some videos about how to use it

javabster commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,105 @@
+---
+
+
+# Use the Qiskit Code Assistant in JupyterLab
```
This intro looks like its for the jupyter lab page not the vscode page?

cbjuan commented:
I’ll stop editing the files unless you ask for it so I don’t interrupt others

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+
+![The bottom status bar is shown, with `granite-8b-qiskit` listed.](/images/guides/qiskit-code-assistant/vscode-selected-model.png)
+
+The first time you use the `granite-8b-qiskit` model a modal opens, listing some major restrictions that you should be aware of when using the model. Clicking `Accept` will accept the disclaimer and enable the model for code generation.
```
```suggestion
The first time you use the `granite-8b-qiskit` model, a modal opens listing some major restrictions that you should be aware of when using the model. Clicking `Accept` will accept the disclaimer and enable the model for code generation.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+![The window that opens upon first use contains restrictions tha must be accepted.](/images/guides/qiskit-code-assistant/eula.png)
+
+### Generate Code
+After accepting the EULA you can open a new file and start coding. To get a code suggestion, type `Ctrl+.` as shown:
```
```suggestion
After accepting the EULA, you can open a new file and start coding. To get a code suggestion, type `Ctrl` + `.` as shown:
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+Press `Tab` to accept the model-generated suggestion or press `ESC` to reject it.
+
+<Admonition type="tip" title="Tips for working with code suggestions">
+- Type `Ctrl+.` again to generate additional suggestions.
```
```suggestion
- Type `Ctrl` + `.` again to generate additional suggestions.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+
+<Admonition type="tip" title="Tips for working with code suggestions">
+- Type `Ctrl+.` again to generate additional suggestions.
+- When a suggestion is shown, you can use `Alt/Option+]` and `Alt/Option+[` to cycle forward and backward, respectively, through the suggestions (if available).
```
```suggestion
- When a suggestion is shown, you can use `Alt/Option` + `]` and `Alt/Option + `[` to cycle forward and backward, respectively, through the suggestions (if available).
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+
+If the extension is installed and enabled, but cannot select a model, verify that your current API Token has been added and the `Qiskit Code Assistant: Url` is properly set.
+
+## Contribute to the JupyterLab extension
+
+The code for this official extension is publicly available and open source. Check it out in [github.](https://github.com/Qiskit/qiskit-code-assistant-jupyterlab)
```
This should be for VSCode, right?

beckykd replied:
```suggestion
The code for this official extension is publicly available and open source. Check it out in [github.](https://github.com/Qiskit/qiskit-code-assistant-vscode)
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+## Next steps
+
+<Admonition type="tip" title="Recommendations">
+    - Check out examples to use the Qiskit Code Assistant for [circuits](https://docs.quantum.ibm.com/guides/circuit-library), [configuring error suppression](https://docs.quantum.ibm.com/guides/configure-error-suppression), and [transpiling with passmanagers.](transpile-with-pass-managers)
```
This won't be love on Monday morning. Should we remove this section for now?

beckykd replied:
Yep

Eric-Arellano reviewed (commented):
(still reviewing)

beckykd commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+
+If the extension is installed and enabled, but cannot select a model, verify that your current API Token has been added and the `Qiskit Code Assistant: Url` is properly set.
+
+## Contribute to the JupyterLab extension
```
```suggestion
## Contribute to the VSCode extension
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+</Admonition>
+
+
+<video title="A user uses the Qiskit Code Assistant in JupyterLab. The user has a comment in the code to get the available backends for their account in IBM Quantum, the Qiskit Code Assistant gives a code suggestion to get the available backends. The user executes the code, gets the available backends in IBM Quantum Platform for the current account. The user ask to the Qiskit Code Assistant to ge ...
```
```suggestion
<video title="A user uses the Qiskit Code Assistant in JupyterLab. The user has a comment in the code to get the available backends for their account in IBM Quantum, the Qiskit Code Assistant gives a code suggestion to get the available backends. The user executes the code, gets the available backends in IBM Quantum Platform for the current account. The user asks the Qiskit Code Assistant to generate the code to get the T1, T2 and other properties for the qubit 1 in the device ibm_brisbane. The assistant gives a code suggestion. The user runs the suggested code and gets the desired output." className="max-w-auto h-auto" controls>
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
````diff
@@ -0,0 +1,133 @@
+
+## Install the JupyterLab extension
+To install the [JupyterLab extension](https://github.com/Qiskit/qiskit-code-assistant-jupyterlab), run the following command from a terminal:
+```sh
````
```suggestion

```sh
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+![Screenshot of the bar on the bottom of JupyterLab.](/images/guides/qiskit-code-assistant/jupyterlab-selected-model.png)
+
+For help working with JupyterLab, refer to the [JupyterLab documentation.](https://jupyterlab.readthedocs.io/en/latest/)
```
Iirc, you prefer this style?

```suggestion
Refer to the [JupyterLab documentation.](https://jupyterlab.readthedocs.io/en/latest/) for help working with JupyterLab
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+## Configure extension settings
+
+It is recommended that you edit the following JupyterLab settings by going Settings > Settings Editor:
```
Not sure if you agree with the arrow suggestion

```suggestion
It is recommended that you edit the following JupyterLab settings by going to Settings -> Settings Editor:
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+- Keyboard shortcuts can be changed from Settings > Settings Editor > Keyboard Shortcuts.
+
+- You can change the IBM Quantum API token to use in the JupyterLab command palette. To do that, type `Alt + Shift + C`, search for `qiskit`, select the **Qiskit Code Assistant: Set IBM Quantum API token** command, and paste in your token.
```
```suggestion
- You can change the IBM Quantum API token to use in the JupyterLab command palette. To do that, type `Alt` + `Shift` + `C`, search for `qiskit`, select the **Qiskit Code Assistant: Set IBM Quantum API token** command, and paste in your token.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+- [Advanced] To change the instance of the Qiskit Code Assistant Service that the extension should use, edit the Qiskit Code Assistant `serviceUrl` setting.
+
+- [Advanced] Keyboard shortcuts can be changed by searching for `completer` in the Keyboard Shortcuts settings (Settings > Settings Editor > Keyboard Shortcuts) and adding new shortcuts for the relevant commands.
```
Feel free to disagree

```suggestion
- [Advanced] Keyboard shortcuts can be changed by searching for `completer` in the Keyboard Shortcuts settings (Settings -> Settings Editor -> Keyboard Shortcuts) and adding new shortcuts for the relevant commands.
```

beckykd replied:
The dash is fine.  In my old tool, typing > added a little arrow, so this was just forgetfulness.

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+Inline autocomplettion can be triggered by using the following key strokes:
+
+- `Alt .` or `Alt \` runs autocomplete at the current cursor location.
```
```suggestion
- `Alt` + `.` or `Alt` + `\` runs autocomplete at the current cursor location.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+Inline autocomplettion can be triggered by using the following key strokes:
+
+- `Alt .` or `Alt \` runs autocomplete at the current cursor location.
+- `Alt [` and `Alt ]` can be used to cycle through the list of suggestions if there are
```
```suggestion
- `Alt` + `[` and `Alt` + `]` can be used to cycle through the list of suggestions if there are
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+- `Alt .` or `Alt \` runs autocomplete at the current cursor location.
+- `Alt [` and `Alt ]` can be used to cycle through the list of suggestions if there are
+  more than one.
+- `Alt Tab` or `Alt END` accepts the suggested code and inserts it at the current
```
```suggestion
- `Alt` + `Tab` or `Alt` + `END` accepts the suggested code and inserts it at the current
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+The service can sometimes take a few seconds to return a suggestion, you can see when the service is working by checking the status bar.
+</Admonition>
+
+Jupyterlab also includes a traditional suggestion context menu. Use `Tab` to run and display the context menu.
```
```suggestion
Jupyterlab also includes a traditional suggestion context menu. Use the `Tab` key to run and display the context menu.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+## Contribute to the JupyterLab extension
+
+The code for this extension is publicly available and open source. Check it out in [github.](https://github.com/Qiskit/qiskit-code-assistant-jupyterlab)
```
```suggestion
The code for this extension is publicly available and open source. Check it out in [GitHub.](https://github.com/Qiskit/qiskit-code-assistant-jupyterlab)
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-jupyterlab.mdx:
```diff
@@ -0,0 +1,133 @@
+
+The code for this extension is publicly available and open source. Check it out in [github.](https://github.com/Qiskit/qiskit-code-assistant-jupyterlab)
+
+## Next steps
```
Ditto if this should be removed for now?

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+
+![A code suggestion is shown.  The suggested code is in grey.](/images/guides/qiskit-code-assistant/vscode-inline-suggestion.png)
+
+Press `Tab` to accept the model-generated suggestion or press `ESC` to reject it.
```
```suggestion
Press the `Tab` key to accept the model-generated suggestion or press the `ESC` key to reject it.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant-vscode.mdx:
```diff
@@ -0,0 +1,100 @@
+</Admonition>
+
+
+<video title="A user uses the Qiskit Code Assistant in Visual Studio Code. The user has previous imports and a function definition called get_random_linear_function with a docstring defining what the function should do. After typing Ctrl+. the Qiskit Code Assistant autocompletes the function with the code to generate a random linear function using Numpy and Qiskit. The video ends with the user ex ...
```
```suggestion
<video title="A user uses the Qiskit Code Assistant in Visual Studio Code. The user has previous imports and a function definition called get_random_linear_function, with a docstring defining what the function should do. After typing the Ctrl and . keys, the Qiskit Code Assistant autocompletes the function with the code to generate a random linear function using Numpy and Qiskit. The video ends with the user executing the code generated and getting a matrix representing a random linear function in the terminal." className="max-w-auto h-auto" controls>
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,76 @@
+
+# Qiskit Code Assistant
+
+The Qiskit Code Assistant aims to make quantum computing more accessible to new Qiskit adopters, and improve the coding experience for current users. It is a generative AI code assistant powered by [watsonx](https://www.ibm.com/products/watsonx-ai). It is trained using millions of text tokens from Qiskit SDK v1.x, years of Qiskit code examples and IBM Quantum&trade; features. The Qiskit Code Assi ...
```
```suggestion
The Qiskit Code Assistant aims to make quantum computing more accessible to new Qiskit adopters and to improve the coding experience for current users. It is a generative AI code assistant powered by [watsonx](https://www.ibm.com/products/watsonx-ai). It is trained using millions of text tokens from Qiskit SDK v1.x, years of Qiskit code examples and IBM Quantum&trade; features. The Qiskit Code Assistant can help your quantum development workflow by offering LLM-generated suggestions based on [IBM Granite 8B Code,](https://www.ibm.com/products/watsonx-ai/foundation-models) which incorporate the latest features and functionalities from IBM&reg;.
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,76 @@
+    - If you have feedback or want to contact the developer team, use the [Qiskit Slack Workspace channel](https://qiskit.enterprise.slack.com/archives/C07LYA6PL83) or the related public GitHub repositories.
+</Admonition>
+
+<video title="A user uses the Qiskit Code Assistant in JupyterLab. The user has a comment in the code to get the available backends for their account in IBM Quantum, the Qiskit Code Assistant gives a code suggestion to get the available backends. The user executes the code, gets the available backends in IBM Quantum Platform for the current account. The user ask to the Qiskit Code Assistant to ge ...
```
```suggestion
<video title="A user uses the Qiskit Code Assistant in JupyterLab. The user has a comment in the code to get the available backends for their account in IBM Quantum. The Qiskit Code Assistant gives a code suggestion to get the available backends. The user executes the code, gets the available backends in IBM Quantum Platform for the current account. The user ask to the Qiskit Code Assistant to generate the code to get the T1, T2 and other properties for the qubit 1 in the device ibm_brisbane. The assistant gives a code suggestion. The user runs the suggested code and gets the desired output." className="max-w-auto h-auto" controls>
```

Eric-Arellano commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,76 @@
+
+## Features
+
+* Accelerates Qiskit code generation by leveraging generative AI based on the `granite-8b-qiskit` model.
+* Use abstract and specific prompts to generate recommendations.
```
The conjugation is inconsistent here. Ditto rest of list

beckykd replied:
yeahhhh - it's ugly because they're not all really the same type of activity.  Let me see if I can fix.

Eric-Arellano commented on docs/guides/qiskit-code-assistant.mdx:
```diff
@@ -0,0 +1,76 @@
+
+<Admonition type="tip" title="Recommendations">
+    - Install and use the official [JupyterLab](/guides/qiskit-code-assistant-jupyterlab) or [VSCode](/guides/qiskit-code-assistant-vscode) extensions.
+    - Follow examples to use the Qiskit Code Assistant for [circuits](https://docs.quantum.ibm.com/guides/circuit-library), [configuring error suppression](https://docs.quantum.ibm.com/guides/configure-error-suppression), and [transpiling with passmanagers.](transpile-with-pass-managers)
```
This should be removed for now

cbjuan commented:
Thank you!! 🙏
